perf_benchmark/benchmark_input_stream.h

```diff
@@ -0,0 +1,121 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+//
+#ifndef PERF_BENCHMARK_BENCHMARK_INPUT_STREAM_H_
+#define PERF_BENCHMARK_BENCHMARK_INPUT_STREAM_H_
+
+#include "grpc_transcoding/transcoder_input_stream.h"
+#include "absl/strings/string_view.h"
+
+namespace google {
+namespace grpc {
+namespace transcoding {
+
+namespace perf_benchmark {
+// An implementation of ZeroCopyInputStream for benchmarking.
+// Subclasses of this should store the entire input message and return pointer
+// to the stored message for each round of Next(). This is useful during
+// benchmark since the same input message will be read multiple times without
+// introducing a large runtime overhead.
+//
+// After each benchmark iteration, Reset() needs to be called.
+class BenchmarkZeroCopyInputStream : public TranscoderInputStream {
+ public:
+  // Pre-Conditions:
```

Comment:
Instead of documenting these pre-conditions, I suggest you add assert to the functions

https://en.cppreference.com/w/cpp/error/assert

Thought i'm not sure if it works in constructor. Nw if it's difficult.

Reply:
You could also just import these common macros from google::protobuf: https://github.com/protocolbuffers/protobuf/blob/a3c8e2deb05186334b7ee8c1174f44802e38b43d/src/google/protobuf/stubs/logging.h